Restyle message boxes and add icon support to them on X11.

eafton opened the pull request:
This patch adds an icon (warning, icon, information) to message boxes on the X11 backend.

eafton commented:
I have decided to redo the positioning/styling code, any opinions?
![Screenshot at 2025-07-06 20-57-38](https://github.com/user-attachments/assets/aa846310-82cf-444a-b186-25f0d52fe50f)
![Screenshot at 2025-07-06 20-57-49](https://github.com/user-attachments/assets/72faf2de-7e36-4799-98c9-93e5720d0ca9)
![Screenshot at 2025-07-06 20-57-53](https://github.com/user-attachments/assets/7b8dd19a-cfdc-4ca8-9202-c223bb5fe8e6)

slouken commented on src/video/x11/SDL_x11messagebox.c:
```diff
@@ -70,20 +78,21 @@ static const SDL_MessageBoxColor g_default_colors[SDL_MESSAGEBOX_COLOR_COUNT] =
     int length;     // Text length
-    int text_width; // Text width
-
+	int text_a;
+	int text_d;
```
```suggestion
    int text_a;
    int text_d;
```

eafton replied:
> Can you please fix your code in this PR to use 4 spaces instead of tabs?

Fixing...

slouken commented on src/video/x11/SDL_x11messagebox.c:
```diff
@@ -70,20 +78,21 @@ static const SDL_MessageBoxColor g_default_colors[SDL_MESSAGEBOX_COLOR_COUNT] =
-    int width;        // Width of this text line
     int length;       // String length of this text line
     const char *text; // Text for this line
+    SDL_Rect rct;
```
```suggestion
    SDL_Rect rect;
```

slouken commented on src/video/x11/SDL_x11messagebox.c:
```diff
@@ -233,6 +278,16 @@ static bool X11_MessageBoxInit(SDL_MessageBoxDataX11 *data, const SDL_MessageBox
+		if (!data->icon_char_font) {
+			data->icon_char_font = X11_XLoadQueryFont(data->display, g_MessageBoxFontLatin1);
+			if (!data->icon_char_font) {
+				return SDL_SetError("Couldn't load icon font %s", g_MessageBoxFontLatin1);
```
Instead of failing, we should probably fall back to the existing behavior of not showing an icon.

slouken commented on src/video/x11/SDL_x11messagebox.c:
```diff
@@ -70,20 +78,21 @@ static const SDL_MessageBoxColor g_default_colors[SDL_MESSAGEBOX_COLOR_COUNT] =
+	int text_a;
+	int text_d;
+	
+    SDL_Rect text_rct;
```
```suggestion
    SDL_Rect text_rect;
```

slouken commented:
> I have decided to redo the positioning/styling code, any opinions? ![Screenshot at 2025-07-06 20-57-38](https://private-user-images.githubusercontent.com/211382996/462946831-aa846310-82cf-444a-b186-25f0d52fe50f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNjE5MjEsIm5iZiI6MTc1MjI2MTYyMSwicGF0aCI6Ii8yMTEzODI5OTYvNDYyOTQ2ODMxLWFhODQ2MzEwLTgyY2YtNDQ0YS1iMTg2LTI1ZjBkNTJmZTUwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQxOTIwMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNDdmZmU3OTEwZWY5MjlmYzQxMTU1MmFiY2UwMWU3ZTcyNmY5Mzc5ZTk1NjJlZmI4MjRhZDAzYjI0YWUzZjdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SdUjN8HthRmYs3sy_t9Wjj475I456AoZXUbz8leNRDE) ![Screenshot at 2025-07-06 20-57-49](https://private-user-images.githubusercontent.com/211382996/462946835-72faf2de-7e36-4799-98c9-93e5720d0ca9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNjE5MjEsIm5iZiI6MTc1MjI2MTYyMSwicGF0aCI6Ii8yMTEzODI5OTYvNDYyOTQ2ODM1LTcyZmFmMmRlLTdlMzYtNDc5OS05OGM5LTkzZTU3MjBkMGNhOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQxOTIwMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02OGUwOTI3NjNjNjRhOWU5YjM2Yjc1NmM2MjVkYjI3NWM3NzM2MzIzNGE3Y2JmODE0YWY1NDBlMTc3Y2RhOTRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aMUAWhuY_8QkvGn2OLBgWHRrkgkr5ghaJYdHGDtfFUo) ![Screenshot at 2025-07-06 20-57-53](https://private-user-images.githubusercontent.com/211382996/462946836-7b8dd19a-cfdc-4ca8-9202-c223bb5fe8e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNjE5MjEsIm5iZiI6MTc1MjI2MTYyMSwicGF0aCI6Ii8yMTEzODI5OTYvNDYyOTQ2ODM2LTdiOGRkMTlhLWNmZGMtNGNhOC05MjAyLWMyMjNiYjVmZThlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQxOTIwMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YTBhMjFlZTk0MThlMTA0YWFmM2Y2ZTExNTExOTY1ZGJjNTM4Y2I4NzMyZDVmMTE4YWIzODhlMTlkZWM5YjA5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Q7yhFjbWYDp1j-7sCUrRdzmiVpLV834WC_shxxIBNsg)

This doesn't add much visual style. Maybe we should use an icon circle background with colors to indicate severity? This more closely matches other platforms.

eafton commented:
> > I have decided to redo the positioning/styling code, any opinions? ![Screenshot at 2025-07-06 20-57-38](https://private-user-images.githubusercontent.com/211382996/462946831-aa846310-82cf-444a-b186-25f0d52fe50f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNjE5MjEsIm5iZiI6MTc1MjI2MTYyMSwicGF0aCI6Ii8yMTEzODI5OTYvNDYyOTQ2ODMxLWFhODQ2MzEwLTgyY2YtNDQ0YS1iMTg2LTI1ZjBkNTJmZTUwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQxOTIwMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNDdmZmU3OTEwZWY5MjlmYzQxMTU1MmFiY2UwMWU3ZTcyNmY5Mzc5ZTk1NjJlZmI4MjRhZDAzYjI0YWUzZjdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SdUjN8HthRmYs3sy_t9Wjj475I456AoZXUbz8leNRDE) ![Screenshot at 2025-07-06 20-57-49](https://private-user-images.githubusercontent.com/211382996/462946835-72faf2de-7e36-4799-98c9-93e5720d0ca9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNjE5MjEsIm5iZiI6MTc1MjI2MTYyMSwicGF0aCI6Ii8yMTEzODI5OTYvNDYyOTQ2ODM1LTcyZmFmMmRlLTdlMzYtNDc5OS05OGM5LTkzZTU3MjBkMGNhOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQxOTIwMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02OGUwOTI3NjNjNjRhOWU5YjM2Yjc1NmM2MjVkYjI3NWM3NzM2MzIzNGE3Y2JmODE0YWY1NDBlMTc3Y2RhOTRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.aMUAWhuY_8QkvGn2OLBgWHRrkgkr5ghaJYdHGDtfFUo) ![Screenshot at 2025-07-06 20-57-53](https://private-user-images.githubusercontent.com/211382996/462946836-7b8dd19a-cfdc-4ca8-9202-c223bb5fe8e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIyNjE5MjEsIm5iZiI6MTc1MjI2MTYyMSwicGF0aCI6Ii8yMTEzODI5OTYvNDYyOTQ2ODM2LTdiOGRkMTlhLWNmZGMtNGNhOC05MjAyLWMyMjNiYjVmZThlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxMVQxOTIwMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YTBhMjFlZTk0MThlMTA0YWFmM2Y2ZTExNTExOTY1ZGJjNTM4Y2I4NzMyZDVmMTE4YWIzODhlMTlkZWM5YjA5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Q7yhFjbWYDp1j-7sCUrRdzmiVpLV834WC_shxxIBNsg)
> 
> This doesn't add much visual style. Maybe we should use an icon circle background with colors to indicate severity? This more closely matches other platforms.

I am looking into it.

eafton commented:
I implemented the suggested changes and switched to a circle background, thoughts?
<img width="369" height="204" alt="screenshot" src="https://github.com/user-attachments/assets/a9c881ed-5631-42ef-bb25-1f4c1129c834" />

eafton commented:
> It doesn't look great to me. Maybe it makes sense to build in actual icons?
<img width="340" height="150" alt="image" src="https://github.com/user-attachments/assets/a38d7691-32ce-4c0b-b88f-f3daad7a671e" />
Is this any better? I think it wouldnt be bad if we used a light color scheme for the widgets.

eafton commented:
@slouken I decided to tweak the palette some more and bevelling to the buttons, thoughts?
<img width="305" height="134" alt="a" src="https://github.com/user-attachments/assets/b27497a3-ff19-4dfb-b8b5-dea654f1eaa3" />
<img width="285" height="149" alt="a2" src="https://github.com/user-attachments/assets/f47abe15-9e1e-4768-99bb-5aaf741d25ba" />
![Uploading a3.png…]()

slouken commented:
This seems fine. @icculus, thoughts?

eafton commented:
> Sorry to barge in, but I have been working on a fork where all of the window and widget code form the messagebox implementation is moved into a built in toolkit so those functions can be reused in a UNIX non-zenity/non-portal dialog implementation, should I work on that or keep this open?

Thoughts?

eafton commented:
> I'll go ahead and merge this. We can tweak the styling for 3.4.0 release. Thanks for your work on this!

No problem, one last thing, should I work on the aforementioned "toolkit"?

slouken commented:
> > I'll go ahead and merge this. We can tweak the styling for 3.4.0 release. Thanks for your work on this!
> 
> @slouken No problem! One last thing, should I work on the "toolkit" mentioned above or not?

It's up to you. I don't think it's something SDL would use, we try to minimize system dependencies where possible.

eafton commented:
> > > I'll go ahead and merge this. We can tweak the styling for 3.4.0 release. Thanks for your work on this!
> > 
> > 
> > @slouken No problem! One last thing, should I work on the "toolkit" mentioned above or not?
> 
> It's up to you. I don't think it's something SDL would use, we try to minimize system dependencies where 



> > > I'll go ahead and merge this. We can tweak the styling for 3.4.0 release. Thanks for your work on this!
> > 
> > 
> > @slouken No problem! One last thing, should I work on the "toolkit" mentioned above or not?
> 
> It's up to you. I don't think it's something SDL would use, we try to minimize system dependencies where possible.

Its all internal to SDL, you can check the branch I sent earlier, I split out the code so it can be reused in a dialog subsystem implementation, I also ported messageboxes to it and it seems to work fine.

slouken commented:
Oh, I see. Feel free to create a PR for this, sure. I don't know whether it makes sense to use for dialogs, but it doesn't hurt to refactor and clean up the messagebox code.
